pkg/controller/controller_broker.go

```diff
@@ -166,27 +167,54 @@ func (c *controller) reconcileServiceBrokerKey(key string) error {
 // reconcileServiceBroker is the control-loop that reconciles a Broker. An
 // error is returned to indicate that the binding has not been fully
 // processed and should be resubmitted at a later time.
-func (c *controller) reconcileServiceBroker(broker *v1alpha1.ServiceBroker) error {
+func (c *controller) reconcileServiceBroker(broker *v1alpha1.ServiceBroker) (bool, error) {
 	glog.V(4).Infof("Processing ServiceBroker %v", broker.Name)
 
+	now := time.Now()
+
 	// If the broker's ready condition is true and the relist interval has not
 	// elapsed, do not reconcile it.
-	if !shouldReconcileServiceBroker(broker, time.Now(), c.brokerRelistInterval) {
+	if !shouldReconcileServiceBroker(broker, now, c.brokerRelistInterval) {
 		glog.V(10).Infof(
 			"Not processing ServiceBroker %v because relist interval has not elapsed since the broker became ready",
 			broker.Name,
 		)
-		return nil
+		return true, nil
+	}
+
+	clone, err := api.Scheme.DeepCopy(broker)
+	if err != nil {
+		return false, err
+	}
+	toUpdate := clone.(*v1alpha1.ServiceBroker)
+
+	// If the reconciler has exceeded the time allowed to retry failed
+	// reconciliations on the broker, then do not reconcile it.
+	if broker.Status.OperationStartTime != nil && !now.Before(broker.Status.OperationStartTime.Time.Add(c.reconciliationRetryDuration)) {
+		s := fmt.Sprintf("Stopping reconciliation retries on ServiceBroker %q because too much time has elapsed", broker.Name)
+		glog.Info(s)
+		c.recorder.Event(broker, api.EventTypeWarning, errorReconciliationRetryTimeoutReason, s)
+		c.setServiceBrokerCondition(toUpdate,
+			v1alpha1.ServiceBrokerConditionFailed,
+			v1alpha1.ConditionTrue,
+			errorReconciliationRetryTimeoutReason,
+			s)
+		c.setServiceBrokerForCompletedChange(toUpdate)
+		c.updateServiceBrokerStatus(toUpdate)
+		return true, stderrors.New(s)
 	}
 
+	c.setServiceBrokerForPendingChange(toUpdate)
+	defer c.updateServiceBrokerStatus(toUpdate)
```

Comment:
not sure why this is defered

Reply:
Same as above, where it seemed better than having so many identical calls to the function littered throughout.